drivers/frequency/adf4371/adf4371.c

```diff
@@ -0,0 +1,820 @@
+}
+
+/**
+ * Set the output frequency for one channel.
+ * @param dev - The device structure.
+ * @param freq - The output frequency.
+ * @param channel - The selected channel.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t adf4371_set_freq(struct adf4371_dev *dev,
+				uint64_t freq,
+				uint32_t channel)
+{
+	uint32_t cp_bleed;
+	uint8_t int_mode = 0;
+	int32_t ret;
+
+	switch (channel) {
+	case ADF4371_CH_RF8:
+	case ADF4371_CH_RFAUX8:
+		if (ADF4371_CHECK_RANGE(freq, OUT_RF8_FREQ))
+			return FAILURE;
+
+		dev->rf_div_sel = 0;
+
+		while (freq < ADF4371_MIN_VCO_FREQ) {
+			freq <<= 1;
+			dev->rf_div_sel++;
+		}
+		break;
+	case ADF4371_CH_RF16:
+		/* ADF4371 RF16 8000...16000 MHz */
+		if (ADF4371_CHECK_RANGE(freq, OUT_RF16_FREQ))
+			return FAILURE;
+
+		freq >>= 1;
+		break;
+	case ADF4371_CH_RF32:
+		/* ADF4371 RF32 16000...32000 MHz */
+		if (ADF4371_CHECK_RANGE(freq, OUT_RF32_FREQ))
+			return FAILURE;
+
+		freq >>= 2;
+		break;
+	default:
+		return FAILURE;
+	}
+
+	adf4371_pll_fract_n_compute(freq, dev->fpfd, &dev->integer, &dev->fract1,
+				    &dev->fract2, &dev->mod2);
+
+	dev->buf[0] = dev->integer >> 8;
+	dev->buf[1] = 0x40; /* REG12 default */
+	dev->buf[2] = 0x00;
+	dev->buf[3] = dev->fract1 & 0xFF;
+	dev->buf[4] = dev->fract1 >> 8;
+	dev->buf[5] = dev->fract1 >> 16;
+	dev->buf[6] = ADF4371_FRAC2WORD_L(dev->fract2 & 0x7F) |
+		      ADF4371_FRAC1WORD(dev->fract1 >> 24);
+	dev->buf[7] = ADF4371_FRAC2WORD_H(dev->fract2 >> 7);
+	dev->buf[8] = dev->mod2 & 0xFF;
+	dev->buf[9] = ADF4371_MOD2WORD(dev->mod2 >> 8);
+
+	ret = adf4371_write_bulk(dev, ADF4371_REG(0x11), dev->buf, 10);
+	if (ret < 0)
+		return ret;
+	/*
+	 * The R counter allows the input reference frequency to be
+	 * divided down to produce the reference clock to the PFD
+	 */
+	ret = adf4371_write(dev, ADF4371_REG(0x1F), dev->ref_div_factor);
+	if (ret < 0)
+		return ret;
+
+	ret = adf4371_update(dev, ADF4371_REG(0x24),
+			     ADF4371_RF_DIV_SEL_MSK,
+			     ADF4371_RF_DIV_SEL(dev->rf_div_sel));
+	if (ret < 0)
+		return ret;
+
+	/*
+	 * The optimum bleed current is set by ((4/N) × ICP)/3.75,
+	 * where ICP is the charge pump current in μA
+	 */
+	cp_bleed = DIV_ROUND_UP(400 * dev->cp_settings.icp, dev->integer * 375);
+	cp_bleed = clamp(cp_bleed, 1U, 255U);
+	ret = adf4371_write(dev, ADF4371_REG(0x26), cp_bleed);
+	if (ret < 0)
+		return ret;
+	/*
+	 * Set to 1 when in INT mode (when FRAC1 = FRAC2 = 0),
+	 * and set to 0 when in FRAC mode.
+	 */
+	if (dev->fract1 == 0 && dev->fract2 == 0)
+		int_mode = 0x01;
+
+	ret = adf4371_write(dev, ADF4371_REG(0x2B), int_mode);
+	if (ret < 0)
+		return ret;
+
+	return adf4371_write(dev, ADF4371_REG(0x10), dev->integer & 0xFF);
+}
+
+/**
+ * Power down one channel.
+ * @param dev - The device structure.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t adf4371_channel_power_down(struct adf4371_dev *dev,
+		uint32_t channel,
```

Comment:
doxygen will complain about 2 undocumented parameters here

Reply:
Right, I've fixed this.